Address WGLC comments (#4)
vasilvv/tls-cross-sni-resumption: draft-ietf-tls-cross-sni-resumption.md

Resumption across server names is negotiated using the TLS flags extension
{{!I-D.draft-ietf-tls-tlsflags}}.  The server MAY send a
resumption_across_names(8) flag in a NewSessionTicket message; the flag is an
assertion by the server that any server for any identity presented in its
certificate would be capable of accepting that ticket.  A client receiving a

> ticket with this flag MAY attempt resumption for any server name listed in the

Reviewer: ```suggestion
ticket with this flag MAY attempt resumption for any server name corresponding to an identity in the
```

Reviewer: (We speak about identities in the cert and then pivot to names, so this just tries to tie them together.)

> server certificate even if the new SNI value does not match the one used in the

Reviewer: ```suggestion
server certificate even if the new server name value does not match the one used in the
```

> server certificate even if the new server name value does not match the one used in the

Reviewer 2: I think that you will need to note somewhere that the client will need to remember the certificate (or the names it contains) in order to be able to perform this resumption.

Author: Added a note to that effect.

> original session.  The flag MUST NOT be used in TLS versions before 1.3.

Reviewer 2: Isn't this "cannot"?  We don't have extensions on NST until 1.3.

Author: That's a good point.  Changed.

4.6.1.  Notably, it still relies on the client ensuring that the server
certificate is valid for the new SNI at the time of session resumption.

> If the origianl server's assertion regarding supporting cross-name resumption

Reviewer: ```suggestion
If the original server's assertion regarding supporting cross-name resumption
```

> turns out to be incorrect, the server receiving a misdirected ticket

Reviewer 2: ```suggestion
turns out to be incorrect, a different server that receives a misdirected ticket
```

> will not be able to decrypt it, and will therefore reject it.  This is secure,

Reviewer 2: ```suggestion
will not be able to decrypt it and will therefore be unable to resume.  This is secure,
```

> as session resumption may be safely rejected for any reason; however, such

Reviewer 2: The reason that this is secure is that the resumption PSK remains available only to the original server.  If the server is not the original, it will be unable to learn anything about the previous session that was not already public knowledge.

What you are saying is that the protocol recovers from this situation gracefully, which is also useful to say, but it is not what ensures that it is secure.

I would use a new sentence rather than a semicolon here before the "however".

Author: Thanks for pointing this out!  "recover gracefully" is what I was looking for here.

misconfiguration will waste tickets stored in the client's cache, as TLS
tickets may be single-use.

When providing the SNI value to the application, TLS 1.3 requires the value

> from the most recent ClientHello to be used ([RFC8446], Section 4.6.1).  If the

Reviewer: ```suggestion
from the most recent ClientHello to be used ({{RFC8446, Section 4.6.1}}).  If the
```

server TLS implementation violates that requirement and instead uses the legacy
behavior of reporting the SNI value of the original session, this can lead to a
confusion attack where the client and the server disagree on the server name

> being used (similar to the attacks described in [DB15]).  It is thus crucial

Reviewer: ```suggestion
being used (similar to the attacks described in {{DB15}}).  It is thus crucial
```

> being used (similar to the attacks described in {{DB15}}).  It is thus crucial

Reviewer 2: "Legacy behaviour" here is something that you are assuming.  It is quite likely that many stacks would need to be updated to report the correct SNI (for both client and server), but you shouldn't label this "legacy".  Simply state that when an implementation supports this feature AND exercises it, then it needs to report the SNI that was used, not the one from the previous connection.  Failure to do so might lead to an attack (as noted).

Author: I think "legacy" here was meant to refer to the fact that certain TLS 1.2 implementations assumed SNI to be a session property, before 1.3 explicitly stated it was per-connection.  But you're right that "legacy" is ambiguous; reworded that.

> for the implementers to ensure that this aspect of SNI processing is handled

Reviewer: ```suggestion
for implementers to ensure that this aspect of SNI processing is handled
```

> correctly before enabling cross-name resumption.

Reviewer 2: "It is thus crucial" might be better replaced with a MUST, lest we mint another RFC 6919 phrase.

Author: That's fair, though this is largely restating an existing MUST in RFC 8446 with a bit of context about what can go wrong if it's violated in this specific scenario.
